package aed3;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.text.Normalizer;
import java.util.regex.Pattern;

public class ParNomeId implements aed3.InterfaceArvoreBMais<ParNomeId> {

  private String nome;
  private int id;
  private short TAMANHO = 30;

  public ParNomeId() throws Exception {
    this("", -1);
  }

  public ParNomeId(String n) throws Exception {
    this(n, -1);
  }

  public ParNomeId(String n, int i) throws Exception {
    if(n.getBytes().length>26)
      throw new Exception("Nome extenso demais. Diminua o número de caracteres.");
    this.nome = n;
    this.id = i;
  }

  @Override
  public ParNomeId clone() {
    try {
      return new ParNomeId(this.nome, this.id);
    } catch (Exception e) {
      System.out.println("Erro na clonagem do objeto ParNomeId");
    }
    return null;
  }

  public short size() {
    return this.TAMANHO;
  }

  public String getNome() {
    return nome;
  }

  public int getId() {
    return id;
  }

  public int compareTo(ParNomeId a) {  
    String str1 = transforma(this.nome);
    String str2 = transforma(a.nome);

    // reduz o tamanho da segunda string para busca
    if(str2.length() > str1.length())
      if(this.id == -1)
        str2 = str2.substring(0, str1.length());
        
    // compara as strings
    if(str1.compareTo(str2)==0)
      if(this.id == -1)
        return 0;
      else
        return this.id - a.id;
    else
      return str1.compareTo(str2);
  }

  public String toString() {
    return this.nome + ";" + String.format("%-3d", this.id);
  }

  public byte[] serialize() throws IOException {
    ByteArrayOutputStream baos = new ByteArrayOutputStream();
    DataOutputStream dos = new DataOutputStream(baos);
    byte[] vb = new byte[26];
    byte[] vbNome = this.nome.getBytes();
    int i=0;
    while(i<vbNome.length && i<26) {
      vb[i] = vbNome[i];
      i++;
    }
    while(i<26) {
      vb[i] = ' ';
      i++;
    }
    dos.write(vb);
    dos.writeInt(this.id);
    return baos.toByteArray();
  }

  public void deserialize(byte[] ba) throws IOException {
    ByteArrayInputStream bais = new ByteArrayInputStream(ba);
    DataInputStream dis = new DataInputStream(bais);
    byte[] vb = new byte[26];
    dis.read(vb);
    this.nome = (new String(vb)).trim();
    this.id = dis.readInt();
  }

  public static String transforma(String str) {
    String nfdNormalizedString = Normalizer.normalize(str, Normalizer.Form.NFD);
    Pattern pattern = Pattern.compile("\\p{InCombiningDiacriticalMarks}+");
    return pattern.matcher(nfdNormalizedString).replaceAll("").toLowerCase();
  }

}
